### ISSUE TITLE
IPv6 support for setNodeAddress()

### ISSUE THREAD

**Issue by pmichali** [kubernetes/kubernetes#44848]
**Is this a request for help?** : No

**What keywords did you search in Kubernetes issues before filing this one?** : node status

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): latest on master

**Environment**:
- **Cloud provider or hardware configuration**: Bare-metal, 3 node config using KubeAdm
- **OS** (e.g. from /etc/os-release): CentOS 7
- **Kernel** (e.g. `uname -a`): Linux devstack-77 3.10.0-514.10.2.el7.x86_64 #1 SMP Fri Mar 3 00:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: KubeAdm
- **Others**:

**What happened**: Scan of code shows that setNodeAddress() in pkg/kubelet/kubelet_node_status.go has logic that handles IPv4 case only.

**What you expected to happen**: Be able to support IPv6 as well.

**How to reproduce it** (as minimally and precisely as possible): Not sure how to reproduce with running system. Will create UT to exercise the IPv6 case.

**Anything else we need to know**: I will work on pushing up a change for this.

**Comment by pmichali** [comment]
In setNodeAddress(), it sets the Legacy Host IP (not sure what this is), and internal IP, based on an IP address that is obtained. For non-cloud providers, it uses the first working of four methods, in order:

- Node IP
- IP from hostname (if IP vs name provided)
- DNS lookup of IP
- IP of network interface that is used for default GW

For the last two items, they are using logic that is designed for IPv4 only. The DNS lookup just does a check of IP and makes sure it is IPv4 and not a loopback. The last one, however, is more complicated.

It calls ChooseHostInterface(), which uses /proc/net/route to all obtain route entries, finds default routes, gets the IPs for the interface, and returns one that is an IPv4 address that is not a loopback, LL multicast, or LL unicast address.

This method is called from several other places (federation, proxiers). Some of the private methods called by ChooseHostInterface may also be called from other places.

I was thinking of four possible ways of attacking this, and would like to hear peoples' thoughts:

1. Create a new ChooseHostInterface**IP**() method, and set of supporting (private) methods that are IPv6 capable. Other callers could switch to this, when supporting IPv6, and the old code removed, once no one is using it.
2. Create a new ChooseHostInterface**IP**() method, that calls the supporting methods, but passes a flag to indicate IPv6 support. Existing ChooseHostInterface() would call the support methods with flag indicating IPv4. Offers some code sharing. Would have the same migration path as the first option.
3. Add a flag to ChooseHostInterface() to indicate IPv6 support. Would require changing other callers to set the flag for IPv4. They could then change it to indicate IPv6, once they are updated. Doesn't seem as clear, passing a flag. Much wider surface area for change. Possibility of 
4. Change all callers of ChooseHostInterface() to be able to support IPv6. Huge surface area to this change - added for completeness.

I was thinking of option #2. Thought? Suggestions? Alternatives?

**Comment by pmichali** [comment]
Two more questions...

The code currently directly accesses /net/proc/route and to do IPv6 we'd need to parse /net/proc/ipv6_route, which has a different format. Is there any reason we cannot use "ip [-6] route", which seems to output in the same format (other than IP addresses themselves)?

To support v6, it means collecting both IPv4 and IPv6 routes (two files, or two commands). My thought was that for option #2 above, I think I can remove all flags in the lower level functions, by having the getRoutes() function only return v4, or v4+v6 routes. I was thinking of having this only returning default routes, instead of all routes and then filtering higher up. I think it'll simplify things. Comments/reactions?

**Comment by pmichali** [comment]
IOW, could I rely on using exec.Command() and provide it with "ip route && ip -6 route | grep default", instead of reading a file?

If so, should I use full path to ip and grep? Will it be consistent across OSes?

**Comment by pmichali** [comment]
Some questions I have on this...

- Do we need/want the ability to enforce an ordering of IP address selection (e.g. IPv4 over IPv6) vs just first global IP found?

- If ordered, do we need/want to apply that for all interfaces (e.g. check all interfaces for IPv4, then check all for IPv6) or on a per interface basis (e.g. check IPv4 and then IPv6 on first interface, then do same on next, etc)?

- If ordered, do we need/want the flexibility (this is not a configuration or API specified option, BTW) in that ordering (e.g. check IPv4 first, or check IPv6 first), or should a fixed precedence be assumed (e.g. always IPv4 first)?

- If ordered, do we want to apply that to both checks - the one using interfaces with default routes, and the one checking all system interfaces?

- Do we need/want the flexibility to exclude an address family (e.g. IPv6 only)?

**Comment by pmichali** [comment]
Just a note... the original code, while specific to IPv4, will look for interfaces that have a default route, and then accept the first global IPv4 address. This may not be the interface from the default route.

**Comment by pmichali** [comment]
I'll push a new patch incorporating some of the feedback from the previous commit and making some decisions regarding the implementation. Specifically...

- Will have a fixed precedence of IPv4 over IPv6 (backwards compatible), for both default route check and system interface check.
- Expecting that operator will have set-up IPv6 expecting to use it, so will alway check for them (after IPv4).
- When checking default routes, will look for IP on same net as default route, rather than any global IP on the interface.

**Comment by pmichali** [comment]
Since pushing this, I see there are some (old) issues discussing removing some of the code in the PR 45116 area, and to make arch. specific versions of ChooseHostInterface().  Not sure of the impact on the discussion here.

**Comment by pmichali** [comment]
I'm going to break this up into multiple commits that are smaller, so they can be reviewed independently (and be smaller - I hope).

**Comment by pmichali** [comment]
Here's the plan I was thinking of for now...

1. Modify the code that obtains the IP by looking at all the system interfaces. I did PR 46044 for that, as it is independent of the rest of the changes.
2. Refactor findMatchingIP() to make sure the IP found on the interface is on same subnet as GW. This is so that dual-stack wouldn't grab the wrong IP (e.g. IPv4 and IPv6 global IPs, but the IPv6 is on the default GW.
3. Obtain default route for IPv6. Need community input as to whether I could use the "ip route/ip -6 route" commands or need to parse the /proc/net/{route,ipv6_route} files.

Please let me know if this seems OK. Also, let me know if I should reuse the existing PR 45116 or close that and use new ones for items #2 and #3.

**Comment by pmichali** [comment]
Some follow-up and questions.

Q1) On item one above, should I create a release note for 46044 saying that it will support IPv6, with a priority on IPv4, when checking the system interfaces as a fallback method for getting the node IP?

Q2) On item two above, rather than checking IPs on the interface and seeing if they are not link-local/loopback/point-to-point, I'm checking that the IP is on the same network as the GW. The intent is to ensure the right IP is selected, when dual-stack. I'm making the assumption that the default route will be for a "globally" visible network and will not have a link-local gateway. Is that an OK assumption?

Q3) @thockin Indicated that for item 3 above, I should use the parsing of files, instead of the "ip route" commands. I intent to create a parser or the ipv6_route command. Should I repurpose PR 445116 (the original PR that had all the changes, but was using ip route commands) for this final step, or should that PR be closed and a new PR for the final step?

Q4-6) With multiple PRs, how do I handle the closing of this issue? Do I use the final commit/PR as the one that "fixes" the issue? What about the other PRs? Do I remove the "fixes" line from them?

Please let me know if I'm moving in the wrong direction on any of this. Trying to feel out the process here.

Thanks!

**Comment by thockin** [comment]
1) If you put it in release notes, people will expect v6 to work - I don't think we're there yet?

2) I don't know - I have repeatedly been surprised by people's weird setups, including some that had no default route.  Can you document this assumption better and think through possible failure modes?

3) I would make a new PR

4) I would make all but the last `xref #1234`, and the last one `fixes #1234`

**Comment by pmichali** [comment]
@thockin Thanks for the responses. Wasn't sure if I needed to include in release notes, some of the tactical changes being made for some of the processing.

I pushed out (new PR) 46138 with info on how the default route was being handled. Have a check that IP for GW is global unicast, as an extra check. The (long) PR comment attempts to explain the changes made to the selection of IP from default route.

I closed 45116. Let me know if I need to do anything more/different to get rid of that PR.

Changed PRs to reference this issue with xref.
